File: product-of-array-except-self.cpp
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> productExceptSelf(vector<int> &nums) {
    int product = 1;
    int length = nums.size();
    vector<int> products(length, 1);

    for (int i = 0; i < length; ++i) {
      products[i] = product;
      product *= nums[i];
    }

    product = 1;
    for (int i = length - 1; i >= 0; --i) {
      products[i] = products[i] * product;
      product *= nums[i];
    }

    return products;
  }
};

File: product-of-array-except-self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        products: List[int] = []
        product = 1

        # prefix product
        for num in nums:
            products.append(product)
            product *= num

        # suffix product
        product = 1
        for i in range(len(nums) - 1, -1, -1):
            products[i] = products[i] * product
            product *= nums[i]

        return products
